js/core/template_manager.js

```diff
@@ -0,0 +1,270 @@
+
+    const widgetElement = $('<div>');
+    const widgetOptions = model.options || {};
+
+    if(DEPRECATED_WIDGET_NAMES.has(widgetName)) {
+        const deprecatedName = widgetName;
+        widgetName = camelize('dx-' + widgetName);
+        log('W0001', 'dxToolbar - "widget" item field', deprecatedName, '16.1', 'Use: ' + widgetName + 'instead');
+    }
+
+    if(parent) {
+        parent._createComponent(widgetElement, widgetName, widgetOptions);
+    } else {
+        widgetElement[widgetName](widgetOptions);
+    }
+
+    return widgetElement;
+});
+const defaultCreateElement = element => new Template(element);
+
+export default class TemplateManager {
+    constructor(createElement, anonymousTemplateName) {
+        this._tempTemplates = [];
+        this._defaultTemplates = {};
+        this._anonymousTemplateName = anonymousTemplateName || ANONYMOUS_TEMPLATE_NAME;
+
+        this._createElement = createElement || defaultCreateElement;
+        this._createTemplateIfNeeded = this._createTemplateIfNeeded.bind(this);
+    }
+
+    static get defaultOptions() {
+        return {
+            integrationOptions: {
+                watchMethod: (fn, callback, options = {}) => {
+                    if(!options.skipImmediate) {
+                        callback(fn());
+                    }
+                    return noop;
+                },
+                templates: { 'dx-polymorph-widget': DX_POLYMORPH_WIDGET_TEMPLATE },
+            }
+        };
+    }
+
+    static _findTemplateByDevice(templates) {
+        const suitableTemplate = findBestMatches(
+            devices.current(),
+            templates,
+            template => getElementOptions(template).dxTemplate
+        )[0];
+
+        templates.forEach((template) => {
+            if(template !== suitableTemplate) {
+                $(template).remove();
+            }
+        });
+
+        return suitableTemplate;
+    }
+
+    static _addOneRenderedCall(template) {
+        const render = template.render.bind(template);
+        return extend({}, template, {
+            render(options) {
+                const templateResult = render(options);
+                options && options.onRendered && options.onRendered();
+                return templateResult;
+            }
+        });
+    }
+
+    static _getNormalizedTemplateArgs(options) {
+        const args = [];
+
+        if('model' in options) {
+            args.push(options.model);
+        }
+        if('index' in options) {
+            args.push(options.index);
+        }
+        args.push(options.container);
+
+        return args;
+    }
+
+    static validateTemplateSource(templateSource) {
+        return typeof templateSource === 'string'
+            ? normalizeTemplateElement(templateSource)
+            : templateSource;
+    }
+
+    static templateKey(templateSource) {
+        return (isRenderer(templateSource) && templateSource[0]) || templateSource;
+    }
+
+    get anonymousTemplateName() {
```

Comment:
It will be better to have method getAnonymousTemplateName because:
1) It will be say by name that it is read only.
2) It will be impossible to make typo error.
3) The getter is magic, method is better to understand.

Reply:
When we use any `method` we think that it calculates something and has some logic inside. When we use any `getter` we know that it don't calculate anything and only returns value of a static field.